src/utils/warehouse/notion/kpi_formatter.rb

```diff
@@ -0,0 +1,132 @@
+
+                kpi_page_id = kpi_record['id']
+                stats_db_id = find_stats_database_id(kpi_page_id, secret: secret)
+
+                format_stats = fetch_and_format_stats_dynamically(stats_db_id, secret: secret, filter_body: filter_body)
+                stats = stats_db_id ? format_stats : []
+
+                formatted_kpi.merge(stats: stats.to_json)
+              end
+            end
+
+            private
+
+            STATS_PROPERTY_EXTRACTORS = {
+              'title' => ->(data) { data.first&.dig('plain_text') },
+              'rich_text' => ->(data) { data.first&.dig('plain_text') },
+              'number' => ->(data) { data },
+              'formula' => lambda do |data|
+                data['number'] if data['type'] == 'number'
+              end
+            }.freeze
+
+            def find_stats_database_id(kpi_page_id, secret:)
+              top_level_blocks = fetch_child_blocks(kpi_page_id, secret: secret)
+              return nil unless top_level_blocks
+
+              year_block = find_year_block(top_level_blocks)
+              return nil unless year_block
+
+              year_block_children = fetch_child_blocks(year_block['id'], secret: secret)
+              return nil unless year_block_children
+
+              db_block = find_database_in_blocks(year_block_children)
+              db_block&.dig('id')
+            end
+
+            ##
+            # Fetches child blocks for a given block ID.
+            def fetch_child_blocks(block_id, secret:)
+              response = notion_request(endpoint: "blocks/#{block_id}/children", method: 'get', secret: secret)
+              response.code == 200 ? response.parsed_response['results'] : nil
+            end
+
+            ##
+            # Finds the year block (Heading 1 with a 4-digit number) in a list of blocks.
+            def find_year_block(blocks)
+              blocks.find do |block|
+                block['type'] == 'heading_1' && block.dig('heading_1', 'rich_text', 0, 'plain_text')&.match?(/\d{4}/)
+              end
+            end
+
+            ##
+            # Finds the child database block in a list of blocks.
+            def find_database_in_blocks(blocks)
+              blocks.find { |block| block['type'] == 'child_database' }
+            end
+
+            ##
+            # Fetches and formats stats dynamically based on the database ID.|
+            def fetch_and_format_stats_dynamically(database_id, secret:, filter_body: {})
+              response = notion_request(endpoint: "databases/#{database_id}/query", secret: secret, body: filter_body)
+              return [] unless response.code == 200
+
+              response.parsed_response['results'].map do |stat_record|
+                properties_hash = {}
+                stat_record['properties'].each do |prop_name, prop_data|
+                  properties_hash[prop_name] = extract_stat_value(prop_data)
+                end
+                properties_hash
+              end
```

Comment:
This can be simplified. The hash type has a method called `transform_values` for these cases.